src/Infiltrator.jl

```diff
@@ -666,7 +675,25 @@ function all_names(m, pred, symbols = Set(Symbol[]), seen = Set(Module[]))
     return symbols
 end
 
-maybe_quote(x) = (isa(x, Expr) || isa(x, Symbol)) ? QuoteNode(x) : x
+function maybe_quote(@nospecialize x)
+  is_ir_construct(x) && return QuoteNode(x)
+  isa(x, Expr) && return QuoteNode(x)
+  isa(x, Symbol) && return QuoteNode(x)
+  return x
+end
+
+let ir_constructs = collect(DataType,
+      filter(map(n->getfield(Core.IR,n), names(Core.IR))) do x
+        @nospecialize x
+        return isa(x, DataType)
+      end)
+  body = Expr(:block)
+  for ty in ir_constructs
+    push!(body.args, :(x isa $ty && return true))
+  end
+  push!(body.args, :(return false))
+  @eval is_ir_construct(@nospecialize x) = $body
+end
```

Comment:
I get why you've written it this way, but I really wish `is_ir_construct(@nospecialize x) = x in ir_constructs` were a feasible solution here...